pkg/controller/migration/convert/core.go

```diff
@@ -198,6 +238,222 @@ func handleCore(c *components, install *operatorv1.Installation) error {
+				helpers.EnsureKubeControllersPodSpecNotNil(install)
+				install.Spec.CalicoKubeControllersDeployment.Spec.MinReadySeconds = &c.kubeControllers.Spec.MinReadySeconds
+			} else {
+				if *minReadySeconds != c.kubeControllers.Spec.MinReadySeconds {
+					return ErrIncompatibleMinReadySeconds(ComponentKubeControllers)
+				}
+			}
+		} else {
+			// minReadySeconds is not set on the component, check if it's set on the installation.
+			if minReadySeconds != nil {
+				return ErrIncompatibleMinReadySeconds(ComponentKubeControllers)
+			}
+		}
+	}
+
+	return nil
+}
+
+func getTolerations(override comp.ReplicatedPodResourceOverrides) []corev1.Toleration {
+	if reflect.ValueOf(override).IsNil() {
+		return nil
+	}
+	return override.GetTolerations()
+}
+
+func areEqualTolerations(tols1 []corev1.Toleration, tols2 []corev1.Toleration) bool {
+	m1 := make(map[corev1.Toleration]int)
+	m2 := make(map[corev1.Toleration]int)
+
+	for _, t := range tols1 {
+		m1[t] += 1
+	}
+	for _, t := range tols2 {
+		m2[t] += 1
+	}
+	for key, v1 := range m1 {
+		if m2[key] != v1 {
+			return false
+		}
+	}
+	for key, v2 := range m2 {
+		if m1[key] != v2 {
+			return false
+		}
+	}
+	return true
+}
+
+// migrateTolerations takes the components and migrates their tolerations to the installation.
+func migrateTolerations(c *components, install *operatorv1.Installation) error {
+	// Handle calico-node
+	installTolerations := getTolerations(install.Spec.CalicoNodeDaemonSet)
+	compTolerations := c.node.Spec.Template.Spec.Tolerations
+
+	// A daemonset/deployment treats nil and empty tolerations the same so do that here too.
+	if len(compTolerations) == 0 {
+		// if the component has no tolerations, then the installation must specify empty, non-nil tolerations.
+		if installTolerations == nil || len(installTolerations) > 0 {
+			return ErrIncompatibleTolerations(ComponentCalicoNode)
+		}
+	} else {
+		// Check if the component tolerations are exactly the same as the default tolerations.
+		// In this case, the following are ok on the installation
+		// - all the default tolerations.
+		// - nil (i.e., not specifying any custom tolerations)
+		//
+		// If the component tolerations are not the default tolerations, ensure they equal those on the installation.
+		if areEqualTolerations(rmeta.TolerateAll, compTolerations) {
+			if installTolerations != nil && !areEqualTolerations(compTolerations, installTolerations) {
+				return ErrIncompatibleTolerations(ComponentCalicoNode)
+			}
+		} else {
+			// if the installation has nil tolerations, copy the component tolerations over.
+			// otherwise compare the tolerations.
+			if installTolerations == nil {
+				helpers.EnsureCalicoNodePodSpecNotNil(install)
+				install.Spec.CalicoNodeDaemonSet.Spec.Template.Spec.Tolerations = compTolerations
+			} else {
+				if !areEqualTolerations(compTolerations, installTolerations) {
+					return ErrIncompatibleTolerations(ComponentCalicoNode)
+				}
+			}
+		}
+	}
```

Comment:
Feels like you could create a function that takes in the `installTolerations, compTolerations, defaultTolerations` and returns the tolerations (that should be on the install), err, then you could use that one function for all node, typha, and kube controllers

Reply:
Addressed in cd45e1a5c8eb0920c69d23f04b64f20fb3bde90a

Probably more refactoring I could do in core.go, I'll take a closer look tomorrow.